src/common/function/src/table/migrate_region.rs

```diff
@@ -0,0 +1,165 @@
+            3 => {
+                let region_ids = cast_u64_vector(&columns[0])?;
+                let from_peers = cast_u64_vector(&columns[1])?;
+                let to_peers = cast_u64_vector(&columns[2])?;
+
+                (region_ids, from_peers, to_peers, None)
+            }
+
+            4 => {
+                let region_ids = cast_u64_vector(&columns[0])?;
+                let from_peers = cast_u64_vector(&columns[1])?;
+                let to_peers = cast_u64_vector(&columns[2])?;
+                let replay_timeouts = cast_u64_vector(&columns[3])?;
+
+                (region_ids, from_peers, to_peers, Some(replay_timeouts))
+            }
+
+            size => {
+                return InvalidFuncArgsSnafu {
+                    err_msg: format!(
+                        "The length of the args is not correct, expect exactly 3 or 4, have: {}",
+                        size
+                    ),
+                }
+                .fail();
+            }
+        };
+
+        std::thread::spawn(move || {
+            let len = region_ids.len();
+            let mut results = StringVectorBuilder::with_capacity(len);
+
+            for index in 0..len {
+                let region_id = region_ids.get(index);
+                let from_peer = from_peers.get(index);
+                let to_peer = to_peers.get(index);
+                let replay_timeout = match &replay_timeouts {
+                    Some(replay_timeouts) => replay_timeouts.get(index),
+                    None => Value::UInt64(DEFAULT_REPLAY_TIMEOUT_SECS),
+                };
+
+                match (region_id, from_peer, to_peer, replay_timeout) {
+                    (
+                        Value::UInt64(region_id),
+                        Value::UInt64(from_peer),
+                        Value::UInt64(to_peer),
+                        Value::UInt64(replay_timeout),
+                    ) => {
+                        let func_ctx = func_ctx.clone();
+
+                        let pid = common_runtime::block_on_read(async move {
+                            func_ctx
+                                .state
+                                .table_mutation_handler
+                                .as_ref()
+                                .context(MissingTableMutationHandlerSnafu)?
+                                .migrate_region(
+                                    region_id,
+                                    from_peer,
+                                    to_peer,
+                                    Duration::from_secs(replay_timeout),
+                                )
+                                .await
+                        })?;
+
+                        results.push(Some(&pid));
+                    }
+                    _ => {
+                        results.push(None);
+                    }
+                }
+            }
+
+            Ok(results.to_vector())
+        })
+        .join()
```

Comment:
If we don't block on a new thread, it will throw an error:

```
2024-02-06T07:45:33.923019Z ERROR do_query:execute_stmt:plan_exec:exec_query_plan:create_physical_plan: common_telemetry::panic_hook: panicked at /Users/dennis/programming/rust/greptimedb/src/common/runtime/src/runtime.rs:81:21:
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks. backtrace=   0: backtrace::backtrace::libunwind::trace
             at /Users/dennis/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/backtrace/libunwind.rs:93:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/dennis/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at /Users/dennis/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/dennis/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/capture.rs:176:9
   3: backtrace::capture::Backtrace::new
             at /Users/dennis/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.69/src/capture.rs:140:22
```

Looks like datafusion has an optimization that will evaluate the `UDF` when creating physical plan:

```
  14: common_runtime::global::block_on_read
             at src/common/runtime/src/global.rs:158:17
  15: <common_function::table::migrate_region::MigrateRegionFunction as common_function::function::Function>::eval
             at src/common/function/src/table/migrate_region.rs:122:35
  16: common_function::scalars::udf::create_udf::{{closure}}
             at src/common/function/src/scalars/udf.rs:66:22
  17: common_query::logical_plan::udf::to_df_scalar_func::{{closure}}
             at src/common/query/src/logical_plan/udf.rs:86:22
  18: <datafusion_physical_expr::scalar_function::ScalarFunctionExpr as datafusion_physical_expr::physical_expr::PhysicalExpr>::evaluate
             at /Users/dennis/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/26e43ac/datafusion/physical-expr/src/scalar_function.rs:154:9
  19: datafusion_optimizer::simplify_expressions::expr_simplifier::ConstEvaluator::evaluate_to_scalar
```

Reply:
This just bypasses the runtime check. It still blocks the runtime thread. Maybe we should adopt async manner for `Function`?

Reply:
Agree. But datafusion doesn't support async UDF currently. cc @waynexia